numpy/lib/function_base.py

```diff
@@ -4393,6 +4460,87 @@ def _quantile_unchecked(a,
         return r
 
 
+def _validate_and_ureduce_weights(a, axis, wgts):
+    """Ensure quantile weights are valid.
+
+    A weight of 1 means a repetition of one for the corresponding value in
+    the data array.  That means the sum of weights along `axis` is the total
+    count of values along that axis, based on which the quantiles are
+    evaluated.
+
+    Weights cannot:
+        * be negative
+        * sum to 0
+    However, they can be
+        * 0, as long as they do not sum to 0
+        * less than 1.  In this case, all weights are re-normalized by
+            the lowest non-zero weight prior to computation.
+
+    Weights will be broadcasted to the shape of a, then reduced as done
+    via _ureduce().
+        * if wgts.shape == a.shape, simply return ureduced wgts.
+        * if scalar, broadcast to shape of a, followed by _ureduce.
+        * for 1-D or N-D array, must make sure that its size matches that of a
+          along the axis dims, so can be reshaped.
+          Example: if a.shape == (2, 3, 4, 5, 6), axis == (1, 3),
+          then weights must be of size 15, and will be reshaped to
+          (1, 3, 1, 5, 1) before broadcasting to the shape of a,
+          followed by _ureduce.
+    """
+    a = asanyarray(a)
+    wgts = asanyarray(wgts)
+
+    if not np.issubdtype(wgts.dtype, float):
+        # convert every element to np.float
+        # raises ValueError if any element fails to convert
+        wgts = wgts.astype(float)
+
+    if np.isnan(wgts).any():
+        raise ValueError("No weight can be NaN.")
+
+    if (wgts < 0).any():
+        raise ValueError("Negative weight not allowed.")
+
+    # dims to reshape to, before broadcast
+    if axis is None:
+        dims = tuple(range(a.ndim))  # all axes
+    else:
+        dims = _nx.normalize_axis_tuple(axis, a.ndim)
+
+    # broadcast weights to the shape of array a.
+    if wgts.shape == a.shape:  # no need to broadcast
+        pass
+    elif wgts.size == 1:  # straight-forwardly broadcastable
+        wgts = np.broadcast_to(wgts, a.shape)
+    elif (wgts.ndim == a.ndim and
+          all([wgts.shape[d] == a.shape[d] or wgts.shape[d] == 1
+               for d in range(a.ndim)])):
```

Comment:
Is there not a function that checks this already? 

If not, I think you can make this more clear (and slightly faster) with the following refactor in `In [9]`:

```
In [9]: %timeit all(d1 == d2 or d1 == 1 for d1, d2 in zip(wgts.shape, a.shape))
490 ns ± 1.06 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [10]: %timeit all([wgts.shape[d] == a.shape[d] or wgts.shape[d] == 1 for d in range(a.ndim)])
656 ns ± 3.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```